### ISSUE TITLE
Get-ItemPropertyValue ignores ErrorAction SilentlyContinue if registry key does not exist

### ISSUE THREAD

**Issue by SwarfegaGit** [PowerShell/PowerShell#5906]
This seems to be a bug in Windows PowerShell 5.1 so I figured I would see if it was resolved in 6.0.0 but it's still there.

Steps to reproduce
------------------

```powershell
Get-ItemPropertyValue -Path 'HKCU:\' -Name xyzzy -ErrorAction SilentlyContinue
```

Expected behavior
-----------------

```none
No error to be displayed
```

Actual behavior
---------------

```powershell
PS C:\Program Files\PowerShell\6.0.0> Get-ItemPropertyValue -Path 'HKCU:\' -Name xyzzy -ErrorAction SilentlyContinue
Get-ItemPropertyValue : Property xyzzy does not exist at path HKEY_CURRENT_USER\.
At line:1 char:1
+ Get-ItemPropertyValue -Path 'HKCU:\' -Name xyzzy -ErrorAction Silentl ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : InvalidArgument: (:) [Get-ItemPropertyValue], PSArgumentException
+ FullyQualifiedErrorId : Argument,Microsoft.PowerShell.Commands.GetItemPropertyValueCommand
```

Environment data
----------------

```powershell
Name                           Value
----                           -----
PSVersion                      6.0.0
PSEdition                      Core
GitCommitId                    v6.0.0
OS                             Microsoft Windows 6.1.7601 S
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by SwarfegaGit** [comment]
As a 5.1 issue and [yesterdays blog](https://blogs.msdn.microsoft.com/powershell/2018/01/24/powershell-core-6-1-roadmap/) mentioning the Consider-WindowsPowerShell51 label is it worth adding that label to this issue?

**Comment by markekraus** [comment]
@SwarfegaGit This is a terminating error. Terminating errors are not suppressed by `-ErrorAction SilentlyContinue`.

**Comment by SwarfegaGit** [comment]
@markekraus so this is a non-terminating error?

```powershell
PS C:\Program Files\PowerShell\6.0.0> Get-ItemProperty -Path 'HKCU:\' -Name xyzzy -ErrorAction SilentlyContinue
PS C:\Program Files\PowerShell\6.0.0> Get-ItemProperty -Path 'HKCU:\' -Name xyzzy
Get-ItemProperty : Property xyzzy does not exist at path HKEY_CURRENT_USER\.
At line:1 char:1
+ Get-ItemProperty -Path 'HKCU:\' -Name xyzzy
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : InvalidArgument: (xyzzy:String) [Get-ItemProperty], PSArgumentException
+ FullyQualifiedErrorId : System.Management.Automation.PSArgumentException,Microsoft.PowerShell.Commands.GetItemPropertyCommand
```
Any reason it should differ?

**Comment by markekraus** [comment]
The difference being, IMO, that `Get-ItemPropertyValue` depends on the existence of the property.  if it didn't terminate when the property didn't exist, the result between the property being unset would be the same as the property not existing at all. With a terminating error you can test for the existence of the property at the same time as getting its value if it has one.

**Comment by KevinMarquette** [comment]
>  if it didn't terminate when the property didn't exist, the result between the property being unset would be the same as the property not existing at all.

If we needed to make that distinction, the error record should give us the info that we need. It feels like there is a terminating error here when there does not need to be. 

As I take a look at this cmdlet, it is inconsistent in how it handles this. I can suppress a bad path, but not a bad property.

    PS> Get-ItemPropertyValue -Path safsdfsfdf -name sdfasfa -ErrorAction SilentlyContinue
    PS> Get-ItemPropertyValue -Path 'HKCU:\bad\to\the\bone' -Name xyzzy -ErrorAction SilentlyContinue
    PS> Get-ItemPropertyValue -Path 'HKCU:\' -Name xyzzy -ErrorAction SilentlyContinue
    Get-ItemPropertyValue : Property xyzzy does not exist at path HKEY_CURRENT_USER\.
    At line:3 char:5
    +     Get-ItemPropertyValue -Path 'HKCU:\' -Name xyzzy -ErrorAction Sil ...

**Comment by mklement0** [comment]
Great analysis, @KevinMarquette, and agreed, @CoreyChaplan:

From https://docs.microsoft.com/en-us/powershell/developer/cmdlet/cmdlet-error-reporting:

> Does the error condition prevent your cmdlet from successfully processing any further input objects? If so, this is a terminating error.

> Is the error condition related to a specific input object or a subset of input objects? If so, this is a nonterminating error.

> Does the cmdlet accept multiple input objects, such that processing may succeed on another input object? If so, this is a nonterminating error.

Therefore, the error at hand should _not_ be a terminating error, given that `Get-ItemPropertyValue` accepts _multiple_ paths via the pipeline.

An aside re your workaround:

~~~
Get-ItemProperty -Path $KeyPath | Select-Object -ExpandProperty $KeyName
~~~

is sufficient - also including `[-Property] $KeyName` is not just unnecessary, it actually _decorates_ the output object with an ETS property of that name.

Depending on your strict-mode settings, even `(Get-ItemProperty -Path $KeyPath).$KeyName` may do.

**Comment by skycommand** [comment]
Thanks for @KevinMarquette and @CoreyChaplan for their observation. Here is a third angle from me:

We already know that the following generates an error:

```` PowerShell
Get-ItemPropertyValue -Path "C:\Windows" -Name "SomethingInvalid" -ErrorAction SilentlyContinue
````

However, the following does not:

```` PowerShell
$ErrorActionPreference="SilentlyContinue"
Get-ItemPropertyValue -Path "C:\Windows" -Name "SomethingInvalid"
$ErrorActionPreference="Continue"
````

I am afraid I don't find the logic of our esteemed @markekraus applicable here.

**Comment by mklement0** [comment]
@skycommand: 

Re contrasting `-ErrorAction SilentlyContinue` with `$ErrorActionPreference="SilentlyContinue"`:

That is a _fundamental_ (problematic) behavioral distinction in PowerShell: 
`-ErrorAction` only acts on _non-terminating_ errors, whereas `$ErrorActionPreference= ...` acts on _all_ errors.

In the case at hand, `$ErrorActionPreference="SilentlyContinue"` therefore manages to suppress the _statement-terminating_ error that `Get-ItemPropertyValue` emits, whereas `-ErrorAction SilentlyContinue` doesn't.

The bottom line is: `Get-ItemPropertyValue` should emit a _non-terminating error_ here, not a (statement-)terminating one.

**Comment by skycommand** [comment]
@mklement0 I'm afraid you're trying to justify one inconsistent bad behavior with another inconsistent bad behavior. In my book, that's two bad behaviors. Aside from that, this cmdlet's behavior is so inconsistent and unpleasant (see the examples above) that anything you say in its defense only serves to hurt your own reputation.

**Comment by mklement0** [comment]
@skycommand:

* Re contrasting `-ErrorAction SilentlyContinue` with `$ErrorActionPreference="SilentlyContinue"`:

I was pointing out that you were referring to  a _fundamental problem_ with PowerShell's error handling and that, as such, it is _incidental to the problem at hand_ - it applies whether or not use of a terminating error is justified.

(If you want to learn about all related warts, read [this](https://github.com/PowerShell/PowerShell-Docs/issues/1583)).

* Re case at hand:

I was pointing out that there is _indeed a problem_ with `Get-ItemPropertyValue` that _should be fixed_: The error that is reported when a value (property) by a given name isn't found should be a _non-terminating error_ (`PSCmdlet.WriteError()`), not a (statement-)terminating one (`PSCmdlet.ThrowTerminatingError()`).

From what I can tell, the only one _not_ in agreement with this in this thread is @markekraus.

> you're trying to justify 
> anything you say in its defense

Let me know if you still feel that something is being justified or defended, or if you spot a flaw in my reasoning above.

>  only serves to hurt your own reputation.

Please refrain from such gratuitous, ad-hominem statements: they add nothing to the discussion and serve only to antagonize.

**Comment by skycommand** [comment]
@mklement0 Really? That's what you wanted to say? Alright. I'm fine with that. I apologize for misunderstanding it. And as a gesture of goodwill, I strike my original reply.

But I emphasize that your original message did not sound like what you are writing now and still does not. You see, you took your time to write what I had already discovered on my own, and reported in a prior post. So, naturally, I'd assume your purpose was something besides informing me, i.e. going from a premise to a conclusion.

**Comment by joakimlemb** [comment]
This bug still exist in 2023, 5 years after initial report...

**Comment by iSazonov** [comment]
@joakimlemb Do you want to pull PR with fix?

**Comment by zaphod4254** [comment]
Another example of why this should be considered a bug and fixed...

The "old" way of reading a registry value, before Get-ItemPropertyValue was introduced,...

(Get-ItemProperty -Path 'HKCU:\').xyzzy

Simply returns null if the registry doesn't exist.  It doesn't even generate an error.

In my opinion then, Get-ItemPropertyValue should function similarly.  At the very least we need to be able to use -ErrorAction SilentlyContinue with Get-ItemPropertyValue.

That said, utilizing the old syntax is an easy workaround, once you know this issue exists.

**Comment by mklement0** [comment]
@zaphod4254 - the proper fix for the issue at hand notwithstanding - see also:

* #14819

**Comment by sappkevin** [comment]
The workaround is to use dot notation for the Get-ItemProperty cmdlet instead of using Get-ItemPropertyValue with -ErrorAction SilentlyContinue 

For example, if I'm checking to see if the HideFirstRunExperience value exist in HKLM:\SOFTWARE\Microsoft\Edge registry key then I would use:
`(Get-ItemProperty -Path "HKLM:\SOFTWARE\Microsoft\Edge\*").HideFirstRunExperience`

Instead of :
`Get-ItemPropertyValue -Path "HKLM:\SOFTWARE\Microsoft\Edge -Name 'HideFirstRunExperience' -ErrorAction SilentlyContinue`

The workaround just gets all of the subitems (Keys, DWORDs, etc) under the registry key and uses dot notation to return the value for a the registry item if it exist or it returns empty/nothing if it does not exist. I hope that helps someone in the future with a similar issue

**Comment by aksarben** [comment]
I heartily agree the error should be non terminating. In my scenario, the error is thrown when I am trying to get a list of registry settings to show the user.

I anticipated the possibility that some properties might not be set. My expectation was that **-ErrorAction Ignore** would make **Get-ItemPropertyValue** return a null object, which I would then convert to “Undefined” for the benefit of the user. But I don't have that option because the current implementation of **Get-ItemPropertyValue** won’t return a null if the property doesn’t exist.

**Comment by bbseal** [comment]
This is still happening and is still 🤡

**Comment by Gijsreyn** [comment]
Would love to see a fix for this one.

**Comment by mrkesu** [comment]
7 years and counting...

**Comment by rgoltz** [comment]
This is just an update to remove the flag added by the bot. Based on the comments above, it's still relevant.

**Comment by zaphod4254** [comment]
I don't really expect this to ever get fixed in PS 5.1, but they should really fix it in the current version.

**Comment by rgoltz** [comment]
This is just an update to remove the flag added by the bot. Based on the comments above, it's still relevant.

**Comment by testsandboxpossum** [comment]
8 years and counting  😜 

Any word on a fix?
